src/awkward/_nplikes/virtual.py

```diff
@@ -66,51 +69,89 @@ def __init__(
         self._dtype = np.dtype(dtype)
         self._array: Sentinel | ArrayLike = UNMATERIALIZED
         self._generator = generator
+        self._shape_generator = shape_generator
 
     @property
     def dtype(self) -> DType:
         return self._dtype
 
     @property
     def shape(self) -> tuple[ShapeItem, ...]:
+        self.get_shape()
         return self._shape
 
+    @property
+    def inner_shape(self) -> tuple[ShapeItem, ...]:
+        if len(self._shape) > 1:
+            return self.shape[1:]
+        return self._shape[1:]
+
     @property
     def ndim(self) -> int:
         return len(self._shape)
 
     @property
     def size(self) -> ShapeItem:
         size: ShapeItem = 1
-        for item in self._shape:
+        for item in self.shape:
             size *= item
         return size
 
     @property
     def nbytes(self) -> ShapeItem:
-        return self.size * self._dtype.itemsize
+        size: ShapeItem = 1
+        for item in self._shape:
+            size *= item
+        return size * self._dtype.itemsize
 
     @property
     def strides(self) -> tuple[ShapeItem, ...]:
         return self.materialize().strides  # type: ignore[attr-defined]
 
+    def get_shape(self) -> None:
+        if any(dim is unknown_length for dim in self._shape):
```

Comment:
Maybe use the property directly?
```suggestion
        if not self.has_known_shape:
```

Reply:
I was kinda following Angus' comment from last time to try to do such things explicitly within the virtual array class. I'll actually remove `has_known_shape` because it's not used anywhere.

Reply:
Ok, then please go ahead and remove it :+1: